common/reactive/src/main/java/io/helidon/common/reactive/Multi.java

```diff
@@ -17,31 +17,222 @@
      */
     default <U> Multi<U> map(Mapper<T, U> mapper) {
-        MultiMappingProcessor<T, U> processor = new MultiMappingProcessor<>(mapper);
+        MultiMapProcessor<T, U> processor = new MultiMapProcessor<>(mapper);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Invoke provided consumer for every item in stream.
+     *
+     * @param consumer consumer to be invoked
+     * @return Multi
+     */
+    default Multi<T> peek(Consumer<T> consumer) {
+        MultiPeekProcessor<T> processor = new MultiPeekProcessor<T>(consumer);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Filter out all duplicates.
+     *
+     * @return Multi
+     */
+    default Multi<T> distinct() {
+        MultiDistinctProcessor<T> processor = new MultiDistinctProcessor<>();
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Filter stream items with provided predicate.
+     *
+     * @param predicate predicate to filter stream with
+     * @return Multi
+     */
+    default Multi<T> filter(Predicate<T> predicate) {
+        MultiFilterProcessor<T> processor = new MultiFilterProcessor<>(predicate);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Take the longest prefix of elements from this stream that satisfy the given predicate.
+     *
+     * @param predicate predicate to filter stream with
+     * @return Multi
+     */
+    default Multi<T> takeWhile(Predicate<T> predicate) {
+        MultiTakeWhileProcessor<T> processor = new MultiTakeWhileProcessor<>(predicate);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Drop the longest prefix of elements from this stream that satisfy the given predicate.
+     *
+     * @param predicate predicate to filter stream with
+     * @return Multi
+     */
+    default Multi<T> dropWhile(Predicate<T> predicate) {
+        MultiDropWhileProcessor<T> processor = new MultiDropWhileProcessor<>(predicate);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Limit stream to allow only specified number of items to pass.
+     *
+     * @param limit with expected number of items to be produced
+     * @return Multi
+     */
+    default Multi<T> limit(long limit) {
+        MultiLimitProcessor<T> processor = new MultiLimitProcessor<>(limit);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Skip first n items, all the others are emitted.
+     *
+     * @param skip number of items to be skipped
+     * @return Multi
+     */
+    default Multi<T> skip(long skip) {
+        MultiSkipProcessor<T> processor = new MultiSkipProcessor<>(skip);
+        this.subscribe(processor);
+        return processor;
+    }
+
+    /**
+     * Coupled processor sends items received to the passed in subscriber, and emits items received from the passed in publisher.
```

Comment:
This method feels quite hard to grasp ; the diagram did not help me understand :(
Also the use of both terms Inlet/upstream and outlet/downstream, we should stick to just one term. IMO upstream and downstream are easier to get around.

I'm not sure I understand exactly what this does. What's the different between this:
```java
Multi.just(List.of("foo", "bar"))
    .coupled(passedInSubscriber, Multi.just(List.of(1, 2)))
    .subscribe(downstreamSubscriber);
```

and this: 

```java
Multi<String> upstream = Multi.just(List.of("foo", "bar"));
upstream.subscribe(passedInSubscriber);

Multi<Integer> passedInPublisher = Multi.just(List.of(1, 2));
passedInPublisher.subscribe(downstreamSubscriber);
```

Reply:
CoupledProcessor only connect signals not the data itself, so I guess best usage is when you want to close parallel stream onError, it may be a little over-complicated but Reactive streams has [coupled](https://download.eclipse.org/microprofile/microprofile-reactive-streams-operators-1.0/apidocs/org/eclipse/microprofile/reactive/streams/operators/ReactiveStreams.html#coupled-org.reactivestreams.Subscriber-org.reactivestreams.Publisher-) and akka has similar thing too [fromSinkAndSourceCoupled](https://doc.akka.io/docs/akka/current/stream/operators/Flow/fromSinkAndSourceCoupled.html). Maybe its better not add it at all, even Clement seems to be quite skeptical about it ["Personally I don’t use the operator, so maybe we should remove it or change it"](https://gitter.im/eclipse/microprofile-reactive#)

Reply:
Given Clement's comment, I would lean toward not exposing this in Multi.

Reply:
removed